test/neo4clj/client_test.clj

```diff
@@ -0,0 +1,207 @@
+
+(t/deftest with-transaction
+  "This test also tests the begin-transaction, commit!, rollback and execute!"
+  (t/testing "Run cypher in a transaction"
+    (test-utils/with-db conn {}
+      (t/is (= (sut/with-transaction conn transaction
+                 (sut/execute! transaction "CREATE (n:TestNode) RETURN n")
+                 (sut/execute! transaction "MATCH (n:TestNode) RETURN n"))
+               [{"n" {:labels [:test-node] :id 0 :ref-id "n" :props {}}}]))))
+  (t/testing "Run cypher in a transaction with rollback on exception"
+    (test-utils/with-db conn {}
+      (t/is (= (do (try
+                     (sut/with-transaction conn transaction
+                       (sut/execute! transaction "CREATE (n:TestNode) RETURN n")
+                       (throw (Exception. "Test Exception")))
+                     (catch Exception e "Caught exception!"))
+                   (sut/execute! conn "MATCH (n:TestNode) RETURN n"))
+               [])))))
+
+(t/deftest execute-read
+  "This test also tests the with-read function"
+  (t/testing "Run read cypher in a Neo4j read-only connection"
+    (test-utils/with-db conn {:initial-data ["CREATE (n:TestNode) RETURN n"]}
+      (t/is (= (sut/execute-read conn "MATCH (n:TestNode) RETURN n")
+               [{"n" {:labels [:test-node] :id 0 :ref-id "n" :props {}}}]))))
+  (t/testing "Run write cypher in a Neo4j read-only connection"
+    (test-utils/with-db conn {}
+      (t/is (thrown? java.lang.Exception
+                     (sut/execute-read conn "CREATE (n:TestNode) RETURN n"))))))
+
+(t/deftest execute-write!
+  "This test also tests the with-write-conn function"
+  (t/testing "Run write cypher without params in a Neo4j write connection"
+    (test-utils/with-db conn {}
+      (t/is (= (sut/execute-write! conn "CREATE (n:TestNode) RETURN n")
+               [{"n" {:labels [:test-node] :id 0 :ref-id "n" :props {}}}]))))
+  (t/testing "Run write cypher with params in a Neo4j write connection"
+    (test-utils/with-db conn {}
+      (t/is (= (sut/execute-write! conn "CREATE (n:TestNode {name: $name}) RETURN n" {:name "Neo"})
+               [{"n" {:labels [:test-node] :id 0 :ref-id "n" :props {:name "Neo"}}}])))))
+
+(t/deftest create-index!
+  (t/testing "Create index on a Neo4j node property"
+    (test-utils/with-db conn {:initial-data ["CREATE (n:TestNode {name: 'Neo'}) RETURN n"]}
+      (t/is (= (map #(select-keys % ["properties" "labelsOrTypes"])
+                    (do (sut/create-index! conn :test-node [:name])
+                        (sut/execute! conn "CALL db.indexes")))
+               [{"properties" ["name"] "labelsOrTypes" ["TestNode"]}])))))
+
+(t/deftest drop-index!
+  (t/testing "Create index on a Neo4j node property"
+    (test-utils/with-db conn {:initial-data ["CREATE INDEX ON :TestNode(name)"]}
+      (t/is (= (do (sut/drop-index! conn :test-node [:name])
+                   (sut/execute! conn "CALL db.indexes")))
+            []))))
```

Comment:
The error in `drop-index!` only happens on my machine because I am using test plugin `pjstadig/humane-test-output`.

I found that it actually highlights an actual bug in the test case. The `[]` should be inside the `=` form:
```clojure
...
      (t/is (= []
               (do (sut/drop-index! conn :test-node [:name])
                   (sut/execute! conn "CALL db.indexes")))))))
```

After that fix, the text case also works with "humane test output".